moshi/src/main/java/com/squareup/moshi/ArrayJsonAdapter.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import kotlin.Throws
+import java.lang.reflect.Array
+import java.lang.reflect.Type
+import okio.IOException
+
+/**
+ * Converts arrays to JSON arrays containing their converted contents. This supports both primitive
+ * and object arrays.
+ */
+internal class ArrayJsonAdapter(private val elementClass: Class<*>, private val elementAdapter: JsonAdapter<Any>
+) : JsonAdapter<Any?>() {
+  @Throws(IOException::class)
+  override fun fromJson(reader: JsonReader): Any {
+    val list = mutableListOf<Any?>()
+    reader.beginArray()
+    while (reader.hasNext()) {
+      list.add(elementAdapter.fromJson(reader))
+    }
+    reader.endArray()
+    val array = Array.newInstance(elementClass, list.size)
+    list.forEachIndexed { i, item ->
+      Array.set(array, i, item)
+    }
```

Comment:
You can use the Array initializer here right? i.e. `Array(size) { index -> ... }`

Reply:
I'm having trouble getting the generics to work here. Do you have any advice?

Reply:
or better yet - why not just use `list.toTypedArray()`?

Reply:
I'm still having trouble because I'm getting an array of Object when I need an array of elementClass

Reply:
I'm not following. You already have a list instance above, you should be able to call toTypedArray() on that no?

Reply:
toTypedArray is creating `Array<Any?>` instead of an Array typed to `elementClass`, and I'm not sure how to create a list with the correct type.

This causes a test failure in CircularAdaptersTest:
```Can not set final [Lcom.squareup.moshi.CircularAdaptersTest$Project; field com.squareup.moshi.CircularAdaptersTest$Team.projects to [Ljava.lang.Object;
java.lang.IllegalArgumentException: Can not set final [Lcom.squareup.moshi.CircularAdaptersTest$Project; field com.squareup.moshi.CircularAdaptersTest$Team.projects to [Ljava.lang.Object;
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at java.base/jdk.internal.reflect.UnsafeQualifiedObjectFieldAccessorImpl.set(UnsafeQualifiedObjectFieldAccessorImpl.java:83)
	at java.base/java.lang.reflect.Field.set(Field.java:799)
	at com.squareup.moshi.ClassJsonAdapter$FieldBinding.read(ClassJsonAdapter.java:246)
	at com.squareup.moshi.ClassJsonAdapter.fromJson(ClassJsonAdapter.java:205)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.Moshi$Lookup.fromJson(Moshi.java:409)
	at com.squareup.moshi.ArrayJsonAdapter.fromJson(ArrayJsonAdapter.kt:32)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.ClassJsonAdapter$FieldBinding.read(ClassJsonAdapter.java:245)
	at com.squareup.moshi.ClassJsonAdapter.fromJson(ClassJsonAdapter.java:205)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.ArrayJsonAdapter.fromJson(ArrayJsonAdapter.kt:32)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.ClassJsonAdapter$FieldBinding.read(ClassJsonAdapter.java:245)
	at com.squareup.moshi.ClassJsonAdapter.fromJson(ClassJsonAdapter.java:205)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.JsonAdapter.fromJson(JsonAdapter.java:70)
	at com.squareup.moshi.CircularAdaptersTest.circularAdapters(CircularAdaptersTest.java:65)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
```

Reply:
Tried it out locally and see what you mean. In that case this is fine for now